x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java

```diff
@@ -281,22 +280,23 @@ default AuthorizationInfo getAuthenticatedUserAuthorizationInfo() {
     }
 
     /**
-     * Used to retrieve index-like resources that the user has access to, for a specific access action type,
+     * Used to retrieve index-like resources that the user has access to, for a specific access action type and selector,
      * at a specific point in time (for a fixed cluster state view).
      * It can also be used to check if a specific resource name is authorized (access to the resource name
      * can be authorized even if it doesn't exist).
      */
     interface AuthorizedIndices {
         /**
-         * Returns all the index-like resource names that are available and accessible for an action type by a user,
+         * Returns all the index-like resource names that are available and accessible for an action type and selector by a user,
          * at a fixed point in time (for a single cluster state view).
+         * The result is cached and subsequent calls to this method are idempotent.
          */
-        Supplier<Set<String>> all();
+        Set<String> all(@Nullable String selector);
```

Comment:
The primary reason this was a supplier (AFAICT) was to ensure that we only do this expensive operation -- most notably here:

https://github.com/elastic/elasticsearch/pull/123986/files#diff-c4c76ddf0f6a4e8b202ec1978c30e11c819ec799615497ceccff93e6c0d1b9dcL326

I've changed the return type to a set instead of a supplier since we can achieve the same guarantee by passing a `Function<String, Set<String>>` instance.